src/cluster_slot.c

```diff
@@ -0,0 +1,244 @@
+ *   * Neither the name of Redis nor the names of its contributors may be used
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "server.h"
+#include "cluster.h"
+#include "cluster_slot.h"
+
+#define DEFAULT_SLOT -1
+#define DEFAULT_STAT 0
+#define UNASSIGNED_SLOT 0
+#define ORDER_BY_KEY_COUNT 1
+#define ORDER_BY_INVALID -1
+
+/* -----------------------------------------------------------------------------
+ * Cluster slot public APIs.
+ * More to be added as part of cluster.c/h refactoring task. See issue #8948.
+ * -------------------------------------------------------------------------- */
+
+int getSlotOrReply(client *c, robj *o) {
+    long long slot;
+    if (getLongLongFromObject(o,&slot) != C_OK ||
+        slot < 0 || slot >= CLUSTER_SLOTS)
+    {
+        addReplyError(c,"Invalid or out of range slot");
+        return -1;
+    }
+    return (int) slot;
+}
+
+/* -----------------------------------------------------------------------------
+ * CLUSTER SLOT-STATS command
+ * -------------------------------------------------------------------------- */
+
+typedef struct sortedSlotStatEntry {
+    int slot;
+    uint64_t stat;
+} sortedSlotStatEntry;
+
+static bool doesSlotBelongToCurrentShard(int slot) {
+    clusterNode *n = server.cluster->myself;
+    return (server.cluster->slots[slot] == n) || (nodeIsSlave(n) && (server.cluster->slots[slot] == n->slaveof));
+}
+
+static void markAssignedSlots(unsigned char *slots) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) slots[slot]++;
+    }
+}
+
+static int countAssignedSlotsFromSlotsArray(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlotsFromCurrentShard() {
+    int count = 0;
+    clusterNode *n = server.cluster->myself;
+    if (nodeIsSlave(n)) {
+        count = n->slaveof->numslots;
+    } else {
+        count = server.cluster->myself->numslots;
+    }
+    return count;
+}
```

Comment:
Seems like this can be a lot more compact.
```suggestion
    clusterNode *myself = server.cluster->myself;
    return nodeIsSlave(server.cluster->myself) ? myself->slaveof->numslots : myself->numslots;
} 
```

Reply:
Also, there might be utility in a function like: `getMasterOfMyshard()`. I see at least one other use case of it. Not a strong opinion though.

Reply:
I like the factoring out idea, call it `getMasterOfThisShard()`.

This should be declared within `cluster.c/h`, as it is a cluster level operation, not at slot level (which would belong to `cluster_slot.c/h`).

So something like;
```
// cluster.c
/* Returns the master node of the shard which the caller node belongs to. */
clusterNode *getMasterOfThisShard() {
    clusterNode *myself = server.cluster->myself;
    return nodeIsSlave(myself) ? myself->slaveof : myself;
}

// cluster_slot.c/h
static int countAssignedSlotsFromThisShard() {
    clusterNode *n = getMasterOfThisShard();
    return n->numslots;
}
```

And to standardize `_ThisShard` naming convention, all functions with `_CurrentShard` can be renamed to `_ThisShard`. Refer to newly proposed names below;
```
countAssignedSlotsFromThisShard();
doesSlotBelongToThisShard();
getMasterOfThisShard();
```

Reply:
I like the factoring out idea, call it `getMasterOfThisShard()`.

This should be declared within `cluster.c/h`, as it is a cluster level operation, not at slot level (which would belong to `cluster_slot.c/h`).

So something like;
```
// cluster.c
/* Returns the master node of the shard which the caller node belongs to. */
clusterNode *getMasterOfThisShard() {
    clusterNode *myself = server.cluster->myself;
    return nodeIsSlave(myself) ? myself->slaveof : myself;
}

// cluster_slot.c/h
static int countAssignedSlotsFromThisShard() {
    clusterNode *n = getMasterOfThisShard();
    return n->numslots;
}
```

The `getMasterOfThisShard` also simplifies the following function `doesSlotBelongToThisShard()`, as the conditional logic on whether the caller node being master or replica is now removed.
```
// before
static int doesSlotBelongToThisShard(int slot) {
    clusterNode *n = server.cluster->myself;
    return (server.cluster->slots[slot] == n) || (nodeIsSlave(n) && (server.cluster->slots[slot] == n->slaveof));
}

// after
static int doesSlotBelongToThisShard(int slot) {
    clusterNode *n = getMasterOfThisShard();
    return server.cluster->slots[slot] == n;
}
```

And to standardize `_ThisShard` naming convention, all functions with `_CurrentShard` can be renamed to `_ThisShard`. Refer to newly proposed names below;
```
countAssignedSlotsFromThisShard();
doesSlotBelongToThisShard();
getMasterOfThisShard();
```